File: frontend/src/utils/recalcCoverage.js
/** Offline recalc path coverage from saved Layer2 dialogues (mirrors eval1/report/recalc_coverage.py). */

export const FLOW_COVERAGE_VIOLATION_THRESHOLD = 0.85;

const INTERRUPTION_NODES = new Set([
  "OBJECTION",
  "F3_RETAIN",
  "FAQ_NORMAL",
  "FAQ_OOB",
  "OBJ_FINAL",
]);

export function parseBotActionSteps(botActionLog) {
  const steps = new Set();
  for (const entry of botActionLog || []) {
    const text = String(entry || "");
    if (!text) continue;
    if (text.includes("opening_line")) steps.add("OPENING");
    if (text.includes(":")) {
      const step = text.split(":").pop()?.trim();
      if (step) steps.add(step);
    }
  }
  return steps;
}

export function inferCoveredNodes(dialogue) {
  const covered = [];
  for (const entry of dialogue?.bot_state_log || []) {
    const sid = String(entry?.current_step_id || "").trim();
    if (sid && !covered.includes(sid)) covered.push(sid);
  }
  return covered;
}

export function mergeEffectiveCoverage(pathNodes, coveredNodes, botActionLog) {
  const visited = new Set((coveredNodes || []).filter(Boolean));
  for (const s of parseBotActionSteps(botActionLog)) visited.add(s);
  if (visited.has("OPENING") && pathNodes.includes("F1")) visited.add("F1");
  if (visited.has("CLOSING") || visited.has("END")) visited.add("CLOSING");
  return visited;
}

export function getApplicablePathNodes(pathNodes, visited) {
  const core = (pathNodes || []).filter((n) => n !== "START" && n !== "END");
  if (!core.length) return [];
  const applicable = [];
  for (const node of core) {
    if (INTERRUPTION_NODES.has(node)) {
      if (visited.has(node)) applicable.push(node);
      continue;
    }
    applicable.push(node);
  }
  return applicable;
}

export function recalcFlowAdherence(dialogue) {
  const pathNodes = dialogue?.path_nodes || [];
  if (!pathNodes.length) return Number(dialogue?.flow_adherence_rate ?? 1);
  const covered = inferCoveredNodes(dialogue);
  const log = dialogue?.bot_state?.bot_action_log || [];
  const visited = mergeEffectiveCoverage(pathNodes, covered, log);
  const applicable = getApplicablePathNodes(pathNodes, visited);
  if (!applicable.length) return 1;
  const hit = applicable.filter((n) => visited.has(n)).length;
  return hit / applicable.length;
}

function runtimeViolations(violations) {
  return (violations || []).filter((v) => v?.violation_type !== "flow_miss");
}

function flowMissViolation(pathId, flow, messageCount) {
  return {
    turn_index: Math.max(1, Math.floor((messageCount || 0) / 2)),
    violation_type: "flow_miss",
    constraint_id: pathId,
    constraint_text: `测试路径 ${pathId} 未按设计节点完整覆盖`,
    bot_utterance: "",
    explanation: `路径节点覆盖率 ${Math.round(flow * 1000) / 10}%（低于 ${FLOW_COVERAGE_VIOLATION_THRESHOLD * 100}% 阈值，修正口径）`,
    deduction: Math.round((1 - flow) * 20 * 100) / 100,
  };
}

function estimateRuleScore({
  violations,
  flow,
  repetitiveBotCount,
  openingLineMatch,
}) {
  let totalDeduction = 0;
  const seen = new Set();
  for (const v of violations) {
    const key = `${v.constraint_id}:${v.turn_index}`;
    if (seen.has(key)) continue;
    seen.add(key);
    let ded = Number(v.deduction || 0);
    if (ded <= 0) {
      const vt = v.violation_type || "";
      if (vt === "hard_boundary") ded = 20;
      else if (vt === "dialogue_length") ded = 7;
      else if (vt === "flow_miss") ded = Math.max(5, (1 - flow) * 20);
      else if (vt === "flow_incomplete") ded = 10;
      else ded = 5;
    }
    totalDeduction += ded;
  }
  if (flow < 0.6) totalDeduction += 10;
  else if (flow < FLOW_COVERAGE_VIOLATION_THRESHOLD) totalDeduction += 5;
  if ((repetitiveBotCount || 0) > 0) {
    totalDeduction += Math.min(15, repetitiveBotCount * 4);
  }
  if (!openingLineMatch) totalDeduction += 8;
  return Math.max(0, Math.round((100 - totalDeduction) * 100) / 100);
}

function gradeForScore(total) {
  if (total >= 90) return "A";
  if (total >= 80) return "B";
  if (total >= 70) return "C";
  if (total >= 60) return "D";
  return "F";
}

function aggregateTotal(ruleScore, llmScore, weightRule = 0.4, weightLlM = 0.6) {
  const total = Math.max(0, Math.min(100, ruleScore * weightRule + llmScore * weightLlM));
  return {
    total_score: Math.round(total * 100) / 100,
    grade: gradeForScore(total),
    score_breakdown: `规则分=${ruleScore}×${weightRule}+LLM分=${llmScore}×${weightLlM} => ${Math.round(total * 100) / 100} (${gradeForScore(total)}) [修正口径]`,
  };
}

/** Patch one report+dialogue merge with recalculated coverage metrics. */
export function recalcReportRecord(report, dialogue, meta = {}) {
  if (!dialogue?.path_nodes?.length) return report;
  const weightRule = Number(meta.weight_rule ?? 0.4);
  const weightLlm = Number(meta.weight_llm ?? 0.6);
  const newFlow = recalcFlowAdherence(dialogue);
  const pathId = report.path_id || dialogue.path_id || "?";

  let violations = runtimeViolations(report.violations);
  if (newFlow < FLOW_COVERAGE_VIOLATION_THRESHOLD) {
    violations = [
      ...violations,
      flowMissViolation(pathId, newFlow, (dialogue.messages || []).length),
    ];
  }

  const ruleScore = estimateRuleScore({
    violations,
    flow: newFlow,
    repetitiveBotCount: dialogue.repetitive_bot_count ?? report.repetitive_bot_count,
    openingLineMatch: dialogue.opening_line_match ?? report.opening_line_match,
  });
  const llmScore = Number(report.llm_score ?? dialogue.llm_score ?? 0);
  const agg = aggregateTotal(ruleScore, llmScore, weightRule, weightLlm);

  return {
    ...report,
    ...dialogue,
    flow_adherence_rate: Math.round(newFlow * 1000) / 1000,
    flow_adherence_rate_legacy: report.flow_adherence_rate,
    violations,
    rule_score: ruleScore,
    total_score: agg.total_score,
    grade: agg.grade,
    score_breakdown: agg.score_breakdown,
    coverage_recalc_applied: true,
  };
}

/** Recalc all reports when layer2 dialogues are present. */
export function recalcReportsFromDialogues(reports, dialogues, meta = {}) {
  const byId = {};
  for (const d of dialogues || []) {
    if (d?.report_id) byId[d.report_id] = d;
  }
  return (reports || []).map((r) => {
    const d = byId[r.report_id];
    return d ? recalcReportRecord(r, d, meta) : r;
  });
}

export function buildRecalcSummary(reports, recalcReports) {
  const oldFm = (reports || []).filter((r) =>
    (r.violations || []).some((v) => v.violation_type === "flow_miss"),
  ).length;
  const newFm = (recalcReports || []).filter((r) =>
    (r.violations || []).some((v) => v.violation_type === "flow_miss"),
  ).length;
  const avg = (list) => {
    if (!list?.length) return null;
    return Math.round((list.reduce((s, r) => s + Number(r.total_score || 0), 0) / list.length) * 10) / 10;
  };
  return {
    flowMissOld: oldFm,
    flowMissNew: newFm,
    avgOld: avg(reports),
    avgNew: avg(recalcReports),
  };
}

File: eval1/report/recalc_coverage.py
"""Recompute path coverage / flow_miss from saved Layer2 dialogues (no re-simulation)."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Dict, List, Optional, Tuple

from eval1.layer2.goal_fsm import FLOW_COVERAGE_VIOLATION_THRESHOLD, GoalFSM
from eval1.layer3.aggregator import Aggregator
from eval1.layer3.rule_judge import RuleJudge
from eval1.layer3.scoring_config import ScoringConfig
from eval1.pipeline.runner import _attach_flow_miss_violation


@dataclass
class _SimplePath:
    nodes: List[str]


@dataclass
class _SimplePlan:
    path: _SimplePath


def infer_covered_nodes(dialogue: Dict[str, Any]) -> List[str]:
    """Rebuild FSM visited steps from bot_state_log."""
    covered: List[str] = []
    for entry in dialogue.get("bot_state_log") or []:
        if not isinstance(entry, dict):
            continue
        sid = str(entry.get("current_step_id") or "").strip()
        if sid and sid not in covered:
            covered.append(sid)
    return covered


def recalc_flow_adherence(dialogue: Dict[str, Any]) -> float:
    path_nodes = list(dialogue.get("path_nodes") or [])
    covered = infer_covered_nodes(dialogue)
    log = list((dialogue.get("bot_state") or {}).get("bot_action_log") or [])
    fsm = GoalFSM.from_path(path_nodes)
    return float(fsm.get_flow_adherence_rate(covered, bot_action_log=log))


def _runtime_violations(violations: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
    return [v for v in violations if str(v.get("violation_type") or "") != "flow_miss"]


def recalc_one(
    report: Dict[str, Any],
    dialogue: Dict[str, Any],
    *,
    weight_rule: float = 0.4,
    weight_llm: float = 0.6,
) -> Dict[str, Any]:
    """Return report fields patched with recalculated coverage + rule score."""
    merged = {**report, **{k: dialogue[k] for k in dialogue if k not in report}}
    new_flow = recalc_flow_adherence(dialogue)
    path_id = str(merged.get("path_id") or dialogue.get("path_id") or "?")

    dialogue_for_rule: Dict[str, Any] = {
        "violations": _runtime_violations(list(merged.get("violations") or [])),
        "flow_adherence_rate": new_flow,
        "hard_violation": bool(merged.get("hard_violation")),
        "repetitive_bot_count": int(merged.get("repetitive_bot_count") or 0),
        "opening_line_match": bool(merged.get("opening_line_match")),
        "bot_state": dict(merged.get("bot_state") or {}),
        "messages": list(merged.get("messages") or []),
    }
    _attach_flow_miss_violation(dialogue_for_rule, path_id)

    plan = _SimplePlan(path=_SimplePath(list(dialogue.get("path_nodes") or [])))
    rule_ret = RuleJudge().score(plan, dialogue_for_rule)

    llm_score = float(merged.get("llm_score") or 0)
    cfg = ScoringConfig(weight_rule=weight_rule, weight_llm=weight_llm)
    agg = Aggregator().aggregate(
        float(rule_ret["rule_score"]),
        llm_score,
        hard_fail=bool(rule_ret.get("hard_violation")),
        scoring=cfg,
    )

    violations: List[Dict[str, Any]] = list(dialogue_for_rule.get("violations") or [])
    for v in rule_ret.get("supplemental_violations") or []:
        if not isinstance(v, dict) or not v:
            continue
        if any(
            str(x.get("constraint_id")) == str(v.get("constraint_id"))
            and str(x.get("violation_type")) == str(v.get("violation_type"))
            for x in violations
        ):
            continue
        violations.append(v)

    out = dict(report)
    out.update(
        {
            "flow_adherence_rate": round(new_flow, 3),
            "flow_adherence_rate_legacy": merged.get("flow_adherence_rate"),
            "violations": violations,
            "rule_score": float(rule_ret["rule_score"]),
            "total_score": float(agg["total_score"]),
            "grade": str(agg["grade"]),
            "score_breakdown": str(agg.get("score_breakdown") or ""),
            "coverage_recalc_applied": True,
        }
    )
    return out


def recalc_eval_payload(
    payload: Dict[str, Any],
    *,
    weight_rule: Optional[float] = None,
    weight_llm: Optional[float] = None,
) -> Tuple[Dict[str, Any], Dict[str, Any]]:
    """Recalc all reports; return (new_payload, summary_delta)."""
    reports = list(payload.get("reports") or [])
    dialogues = {
        str(d.get("report_id")): d
        for d in (payload.get("layer2") or {}).get("dialogues") or []
        if d.get("report_id")
    }
    meta = (payload.get("summary") or {}).get("meta") or payload.get("meta") or {}
    wr = float(weight_rule if weight_rule is not None else meta.get("weight_rule") or 0.4)
    wl = float(weight_llm if weight_llm is not None else meta.get("weight_llm") or 0.6)

    old_flow_miss = 0
    new_flow_miss = 0
    recalc_reports: List[Dict[str, Any]] = []

    for r in reports:
        rid = str(r.get("report_id") or "")
        d = dialogues.get(rid)
        if not d or not d.get("path_nodes"):
            recalc_reports.append(r)
            continue
        if any(v.get("violation_type") == "flow_miss" for v in (r.get("violations") or [])):
            old_flow_miss += 1
        patched = recalc_one(r, d, weight_rule=wr, weight_llm=wl)
        if any(v.get("violation_type") == "flow_miss" for v in (patched.get("violations") or [])):
            new_flow_miss += 1
        recalc_reports.append(patched)

    new_payload = dict(payload)
    new_payload["reports"] = recalc_reports

    # Resummary
    count = len(recalc_reports)
    avg = sum(float(x.get("total_score") or 0) for x in recalc_reports) / count if count else 0
    grades: Dict[str, int] = {}
    dim_sum: Dict[str, float] = {}
    dim_n = 0
    for x in recalc_reports:
        g = str(x.get("grade") or "?")
        grades[g] = grades.get(g, 0) + 1
        ds = x.get("dimension_scores") or {}
        if ds:
            dim_n += 1
            for k, v in ds.items():
                dim_sum[k] = dim_sum.get(k, 0) + float(v)

    summary = dict(payload.get("summary") or {})
    summary["count"] = count
    summary["average_score"] = round(avg, 2)
    summary["grade_distribution"] = grades
    if dim_n:
        summary["dimension_averages"] = {
            k: round(v / dim_n, 2) for k, v in dim_sum.items()
        }
    summary["coverage_recalc"] = True
    new_payload["summary"] = summary

    delta = {
        "cases": count,
        "flow_miss_old": old_flow_miss,
        "flow_miss_new": new_flow_miss,
        "average_score_old": payload.get("summary", {}).get("average_score"),
        "average_score_new": round(avg, 2),
        "grade_distribution_new": grades,
    }
    return new_payload, delta
